## IMDB TMDB Movie Dataset Data Processing

### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly
import plotly.express as px
import plotly.graph_objects as go
import tensorflow as tf
from tensorflow import keras
import ast
import datetime
import sklearn
import scipy
from tqdm.notebook import tqdm
tqdm.pandas()

In [2]:
%%html
<style>
.cell-output-ipywidget-background {
    background-color: transparent !important;
}
:root {
    --jp-widgets-color: var(--vscode-editor-foreground);
    --jp-widgets-font-size: var(--vscode-editor-font-size);
}  

In [3]:
df  = pd.DataFrame(pd.read_csv('../data/IMDB TMDB Movie Metadata Big Dataset (1M).csv'))
df.head()

id            title  vote_average  vote_count    status release_date  \
0   27205        Inception         8.364       34495  Released   2010-07-15   
1  157336     Interstellar         8.417       32571  Released   2014-11-05   
2     155  The Dark Knight         8.512       30619  Released   2008-07-16   
3   19995           Avatar         7.573       29815  Released   2009-12-15   
4   24428     The Avengers         7.710       29166  Released   2012-04-25   

      revenue  runtime  adult                     backdrop_path  ...  \
0   825532764      148  False  /8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg  ...   
1   701729206      169  False  /pbrkL804c8yAv3zBZR4QPEafpAR.jpg  ...   
2  1004558444      152  False  /nMKdUUepR0i5zn0y1T4CsSB5chy.jpg  ...   
3  2923706026      162  False  /vL5LR6WdxWPjLPFRLe133jXWsh5.jpg  ...   
4  1518815515      143  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg  ...   

                Star3               Star4             Writer  \
0         Elliot Page        Ken Watanabe  Christopher Nolan   
1    Jessica Chastain       Mackenzie Foy     Jonathan Nolan   
2       Aaron Eckhart       Michael Caine     Jonathan Nolan   
3    Sigourney Weaver  Michelle Rodriguez      James Cameron   
4  Scarlett Johansson       Jeremy Renner      Sydney Newman   

  Director_of_Photography                                          Producers  \
0           Wally Pfister  Thomas Tull, Christopher Nolan, Chris Brigham,...   
1       Hoyte van Hoytema  Jake Myers, Emma Thomas, Jordan Goldberg, Thom...   
2           Wally Pfister  Kevin De La Noy, Thomas Tull, Christopher Nola...   
3             Mauro Fiore  Jon Landau, Colin Wilson, Laeta Kalogridis, Ja...   
4                     NaN                                                NaN   

                     Music_Composer  \
0                       Hans Zimmer   
1                       Hans Zimmer   
2  Hans Zimmer, James Newton Howard   
3                      James Horner   
4                               NaN   

                                         genres_list  \
0         ['Action', 'Science Fiction', 'Adventure']   
1          ['Adventure', 'Drama', 'Science Fiction']   
2           ['Drama', 'Action', 'Crime', 'Thriller']   
3  ['Action', 'Adventure', 'Fantasy', 'Science Fi...   
4         ['Science Fiction', 'Action', 'Adventure']   

                                           Cast_list overview_sentiment  \
0  ['Tim Kelleher', 'Silvie Laguna', 'Natasha Bea...          -0.011111   
1  ['Jeff Hephner', 'William Devane', 'Elyes Gabe...           0.045455   
2  ['Tommy Lister Jr.', 'Edison Chen', 'Beatrice ...           0.025000   
3  ['Carvon Futrell', 'Joel David Moore', 'Jon Cu...           0.041667   
4  ['Haneyuri', 'Nako Mizusawa', 'Marin', 'Rikako...           0.025000   

                               all_combined_keywords  
0  ['s', 'philosophy', 'skilled', 'kidnapping', '...  
1  ['thoughtful', 'use', 'scientist', 'quantum me...  
2  ['reign', 'harvey', 'proves', 'partnership', '...  
3  ['space war', 'civilization', 'society', 'mari...  
4  ['s', 'world', 'begins', 'e', 'global safety',...  

[5 rows x 42 columns]

In [4]:
df.shape

(1072255, 42)

In [5]:
df.iloc[:,:20].head()

id            title  vote_average  vote_count    status release_date  \
0   27205        Inception         8.364       34495  Released   2010-07-15   
1  157336     Interstellar         8.417       32571  Released   2014-11-05   
2     155  The Dark Knight         8.512       30619  Released   2008-07-16   
3   19995           Avatar         7.573       29815  Released   2009-12-15   
4   24428     The Avengers         7.710       29166  Released   2012-04-25   

      revenue  runtime  adult                     backdrop_path     budget  \
0   825532764      148  False  /8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg  160000000   
1   701729206      169  False  /pbrkL804c8yAv3zBZR4QPEafpAR.jpg  165000000   
2  1004558444      152  False  /nMKdUUepR0i5zn0y1T4CsSB5chy.jpg  185000000   
3  2923706026      162  False  /vL5LR6WdxWPjLPFRLe133jXWsh5.jpg  237000000   
4  1518815515      143  False  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg  220000000   

                                         homepage    imdb_id  \
0     https://www.warnerbros.com/movies/inception  tt1375666   
1               http://www.interstellarmovie.net/  tt0816692   
2  https://www.warnerbros.com/movies/dark-knight/  tt0468569   
3            https://www.avatar.com/movies/avatar  tt0499549   
4      https://www.marvel.com/movies/the-avengers  tt0848228   

  original_language   original_title  \
0                en        Inception   
1                en     Interstellar   
2                en  The Dark Knight   
3                en           Avatar   
4                en     The Avengers   

                                            overview  popularity  \
0  Cobb, a skilled thief who commits corporate es...      83.952   
1  The adventures of a group of explorers who mak...     140.241   
2  Batman raises the stakes in his war on crime. ...     130.643   
3  In the 22nd century, a paraplegic Marine is di...      79.932   
4  When an unexpected enemy emerges and threatens...      98.082   

                        poster_path  \
0  /oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg   
1  /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg   
2  /qJ2tW6WMUDux911r6m7haRef0WH.jpg   
3  /kyeqWdyUXW608qlYkRqosgbbJyK.jpg   
4   /RYMX2wcKCBAr24UyPD7xwmjaTn.jpg   

                                             tagline  \
0               Your mind is the scene of the crime.   
1  Mankind was born on Earth. It was never meant ...   
2                  Welcome to a world without rules.   
3                        Enter the world of Pandora.   
4                            Some assembly required.   

                                production_companies  
0  Legendary Pictures, Syncopy, Warner Bros. Pict...  
1  Legendary Pictures, Syncopy, Lynda Obst Produc...  
2  DC Comics, Legendary Pictures, Syncopy, Isobel...  
3  Dune Entertainment, Lightstorm Entertainment, ...  
4                                     Marvel Studios

In [6]:
df.iloc[:,20:40].head()

production_countries  \
0  United Kingdom, United States of America   
1  United Kingdom, United States of America   
2  United Kingdom, United States of America   
3  United States of America, United Kingdom   
4                  United States of America   

                     spoken_languages  \
0  English, French, Japanese, Swahili   
1                             English   
2                   English, Mandarin   
3                    English, Spanish   
4             English, Hindi, Russian   

                                            keywords  release_year  \
0  ['rescue', 'mission', 'dream', 'airplane', 'pa...        2010.0   
1  ['rescue', 'future', 'spacecraft', 'race again...        2014.0   
2  ['joker', 'sadism', 'chaos', 'secret identity'...        2008.0   
3  ['future', 'society', 'culture clash', 'space ...        2009.0   
4  ['new york city', 'superhero', 'shield', 'base...        2012.0   

            Director  AverageRating  \
0  Christopher Nolan            8.8   
1  Christopher Nolan            8.7   
2  Christopher Nolan            9.0   
3      James Cameron            7.9   
4        Joss Whedon            8.0   

                                         Poster_Link Certificate  IMDB_Rating  \
0  https://m.media-amazon.com/images/M/MV5BMjAxMz...          UA          8.8   
1  https://m.media-amazon.com/images/M/MV5BZjdkOT...          UA          8.6   
2  https://m.media-amazon.com/images/M/MV5BMTMxNT...          UA          9.0   
3  https://m.media-amazon.com/images/M/MV5BMTYwOT...          UA          7.8   
4  https://m.media-amazon.com/images/M/MV5BNDYxNj...          UA          8.0   

   Meta_score                Star1                 Star2               Star3  \
0        74.0    Leonardo DiCaprio  Joseph Gordon-Levitt         Elliot Page   
1        74.0  Matthew McConaughey         Anne Hathaway    Jessica Chastain   
2        84.0       Christian Bale          Heath Ledger       Aaron Eckhart   
3        83.0      Sam Worthington           Zoe Saldana    Sigourney Weaver   
4        69.0    Robert Downey Jr.           Chris Evans  Scarlett Johansson   

                Star4             Writer Director_of_Photography  \
0        Ken Watanabe  Christopher Nolan           Wally Pfister   
1       Mackenzie Foy     Jonathan Nolan       Hoyte van Hoytema   
2       Michael Caine     Jonathan Nolan           Wally Pfister   
3  Michelle Rodriguez      James Cameron             Mauro Fiore   
4       Jeremy Renner      Sydney Newman                     NaN   

                                           Producers  \
0  Thomas Tull, Christopher Nolan, Chris Brigham,...   
1  Jake Myers, Emma Thomas, Jordan Goldberg, Thom...   
2  Kevin De La Noy, Thomas Tull, Christopher Nola...   
3  Jon Landau, Colin Wilson, Laeta Kalogridis, Ja...   
4                                                NaN   

                     Music_Composer  \
0                       Hans Zimmer   
1                       Hans Zimmer   
2  Hans Zimmer, James Newton Howard   
3                      James Horner   
4                               NaN   

                                         genres_list  \
0         ['Action', 'Science Fiction', 'Adventure']   
1          ['Adventure', 'Drama', 'Science Fiction']   
2           ['Drama', 'Action', 'Crime', 'Thriller']   
3  ['Action', 'Adventure', 'Fantasy', 'Science Fi...   
4         ['Science Fiction', 'Action', 'Adventure']   

                                           Cast_list  
0  ['Tim Kelleher', 'Silvie Laguna', 'Natasha Bea...  
1  ['Jeff Hephner', 'William Devane', 'Elyes Gabe...  
2  ['Tommy Lister Jr.', 'Edison Chen', 'Beatrice ...  
3  ['Carvon Futrell', 'Joel David Moore', 'Jon Cu...  
4  ['Haneyuri', 'Nako Mizusawa', 'Marin', 'Rikako...

In [7]:
df.iloc[:,40:].head()

overview_sentiment                              all_combined_keywords
0           -0.011111  ['s', 'philosophy', 'skilled', 'kidnapping', '...
1            0.045455  ['thoughtful', 'use', 'scientist', 'quantum me...
2            0.025000  ['reign', 'harvey', 'proves', 'partnership', '...
3            0.041667  ['space war', 'civilization', 'society', 'mari...
4            0.025000  ['s', 'world', 'begins', 'e', 'global safety',...

In [8]:
df.dtypes

id                           int64
title                       object
vote_average               float64
vote_count                   int64
status                      object
release_date                object
revenue                      int64
runtime                      int64
adult                         bool
backdrop_path               object
budget                       int64
homepage                    object
imdb_id                     object
original_language           object
original_title              object
overview                    object
popularity                 float64
poster_path                 object
tagline                     object
production_companies        object
production_countries        object
spoken_languages            object
keywords                    object
release_year               float64
Director                    object
AverageRating              float64
Poster_Link                 object
Certificate                 object
IMDB_Rating         

### Type Conversion

In [9]:
type(df['release_date'][0])

str

`release_date` is in `str` format. Let's change that to its original format.

In [10]:
df['release_date'] = pd.to_datetime(df['release_date'], errors='coerce')
print(type(df['release_date'][0]))
print(df['release_date'].head())

<class 'pandas._libs.tslibs.timestamps.Timestamp'>
0   2010-07-15
1   2014-11-05
2   2008-07-16
3   2009-12-15
4   2012-04-25
Name: release_date, dtype: datetime64[ns]


In [11]:
array_df = df[['keywords', 'genres_list', 'Cast_list', 'all_combined_keywords']]
array_df.head()

keywords  \
0  ['rescue', 'mission', 'dream', 'airplane', 'pa...   
1  ['rescue', 'future', 'spacecraft', 'race again...   
2  ['joker', 'sadism', 'chaos', 'secret identity'...   
3  ['future', 'society', 'culture clash', 'space ...   
4  ['new york city', 'superhero', 'shield', 'base...   

                                         genres_list  \
0         ['Action', 'Science Fiction', 'Adventure']   
1          ['Adventure', 'Drama', 'Science Fiction']   
2           ['Drama', 'Action', 'Crime', 'Thriller']   
3  ['Action', 'Adventure', 'Fantasy', 'Science Fi...   
4         ['Science Fiction', 'Action', 'Adventure']   

                                           Cast_list  \
0  ['Tim Kelleher', 'Silvie Laguna', 'Natasha Bea...   
1  ['Jeff Hephner', 'William Devane', 'Elyes Gabe...   
2  ['Tommy Lister Jr.', 'Edison Chen', 'Beatrice ...   
3  ['Carvon Futrell', 'Joel David Moore', 'Jon Cu...   
4  ['Haneyuri', 'Nako Mizusawa', 'Marin', 'Rikako...   

                               all_combined_keywords  
0  ['s', 'philosophy', 'skilled', 'kidnapping', '...  
1  ['thoughtful', 'use', 'scientist', 'quantum me...  
2  ['reign', 'harvey', 'proves', 'partnership', '...  
3  ['space war', 'civilization', 'society', 'mari...  
4  ['s', 'world', 'begins', 'e', 'global safety',...

In [12]:
array_df.iloc[0].progress_apply(type)

  0%|          | 0/4 [00:00<?, ?it/s]

keywords                 <class 'str'>
genres_list              <class 'str'>
Cast_list                <class 'str'>
all_combined_keywords    <class 'str'>
Name: 0, dtype: object

The `array` are in `str` format. So we will need to change them back.

In [13]:
array_df = array_df.applymap(ast.literal_eval)
array_df.head()

/var/folders/1v/dvtq9m8d48z659z8qvy5rxdh0000gn/T/ipykernel_1863/1995840605.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  array_df = array_df.applymap(ast.literal_eval)


keywords  \
0  [rescue, mission, dream, airplane, paris, fran...   
1  [rescue, future, spacecraft, race against time...   
2  [joker, sadism, chaos, secret identity, crime ...   
3  [future, society, culture clash, space travel,...   
4  [new york city, superhero, shield, based on co...   

                                     genres_list  \
0           [Action, Science Fiction, Adventure]   
1            [Adventure, Drama, Science Fiction]   
2               [Drama, Action, Crime, Thriller]   
3  [Action, Adventure, Fantasy, Science Fiction]   
4           [Science Fiction, Action, Adventure]   

                                           Cast_list  \
0  [Tim Kelleher, Silvie Laguna, Natasha Beaumont...   
1  [Jeff Hephner, William Devane, Elyes Gabel, To...   
2  [Tommy Lister Jr., Edison Chen, Beatrice Rosen...   
3  [Carvon Futrell, Joel David Moore, Jon Curry, ...   
4  [Haneyuri, Nako Mizusawa, Marin, Rikako Sakata...   

                               all_combined_keywords  
0  [s, philosophy, skilled, kidnapping, manipulat...  
1  [thoughtful, use, scientist, quantum mechanics...  
2  [reign, harvey, proves, partnership, batman, n...  
3  [space war, civilization, society, marine, mis...  
4  [s, world, begins, e, global safety, recruitme...

In [14]:
for attribute in tqdm(list(array_df.keys())):
    df[attribute] = array_df[attribute]

print(array_df.iloc[0].apply(type))
df[['keywords', 'genres_list', 'Cast_list', 'all_combined_keywords']].head()

  0%|          | 0/4 [00:00<?, ?it/s]

keywords                 <class 'list'>
genres_list              <class 'list'>
Cast_list                <class 'list'>
all_combined_keywords    <class 'list'>
Name: 0, dtype: object


keywords  \
0  [rescue, mission, dream, airplane, paris, fran...   
1  [rescue, future, spacecraft, race against time...   
2  [joker, sadism, chaos, secret identity, crime ...   
3  [future, society, culture clash, space travel,...   
4  [new york city, superhero, shield, based on co...   

                                     genres_list  \
0           [Action, Science Fiction, Adventure]   
1            [Adventure, Drama, Science Fiction]   
2               [Drama, Action, Crime, Thriller]   
3  [Action, Adventure, Fantasy, Science Fiction]   
4           [Science Fiction, Action, Adventure]   

                                           Cast_list  \
0  [Tim Kelleher, Silvie Laguna, Natasha Beaumont...   
1  [Jeff Hephner, William Devane, Elyes Gabel, To...   
2  [Tommy Lister Jr., Edison Chen, Beatrice Rosen...   
3  [Carvon Futrell, Joel David Moore, Jon Curry, ...   
4  [Haneyuri, Nako Mizusawa, Marin, Rikako Sakata...   

                               all_combined_keywords  
0  [s, philosophy, skilled, kidnapping, manipulat...  
1  [thoughtful, use, scientist, quantum mechanics...  
2  [reign, harvey, proves, partnership, batman, n...  
3  [space war, civilization, society, marine, mis...  
4  [s, world, begins, e, global safety, recruitme...

### Data Cleaning

In [15]:
df.isnull().sum()

id                               0
title                            0
vote_average                     0
vote_count                       0
status                           0
release_date                150556
revenue                          0
runtime                          0
adult                            0
backdrop_path               779429
budget                           0
homepage                    956736
imdb_id                     482891
original_language                0
original_title                   0
overview                    202181
popularity                       0
poster_path                 317838
tagline                     919472
production_companies        575683
production_countries        456376
spoken_languages            440257
keywords                         0
release_year                150556
Director                         0
AverageRating              1060504
Poster_Link                1069485
Certificate                1042420
IMDB_Rating         

IMDB Dataset Attribute Analysis

Potentially Unnecessary Attributes

1. AverageRating (1,060,504 nulls)
- **Explanation:** This attribute is almost entirely null. It's redundant with `vote_average` and `IMDB_Rating`, which have no null values.

2. Poster_Link (1,069,485 nulls)
- **Explanation:** Nearly all entries are null. The `poster_path` attribute likely serves a similar purpose with fewer null values.

3. Certificate (1,042,420 nulls)
- **Explanation:** Mostly null. If certification information is crucial, it might be better to source it from a more complete dataset.

4. IMDB_Rating (1,041,943 nulls)
- **Explanation:** Largely null and likely redundant with `vote_average`.

5. Meta_score (1,069,924 nulls)
- **Explanation:** Almost entirely null. If Metacritic scores are desired, a more complete source should be used.

6. Music_Composer (916,779 nulls)
- **Explanation:** Largely null. This information might be included in the `Producers` or `Cast_list` attributes.

7. Director_of_Photography (735,402 nulls)
- **Explanation:** Mostly null. This level of detail might not be necessary for general analysis.

8. homepage (956,736 nulls)
- **Explanation:** Mostly null and not crucial for most analyses.

9. tagline (919,472 nulls)
- **Explanation:** Largely null and often not essential for analysis. 

8. Producers (646,947 nulls)
- **Explanation:** Mostly null.

9. Writer (425,675 nulls)
- **Explanation:** While it has fewer nulls than some other attributes, it's still mostly empty and might be redundant with information in Cast_list or overview.

<br>

Attributes to Consider Keeping Despite High Null Counts

1. production_countries (456,376 nulls)
- While it has many nulls, this information could be valuable for regional analysis.

2. spoken_languages (440,257 nulls)
- This could be important for language-based analysis, despite the high null count.

3. imdb_id (482,891 nulls)
- This could be useful for linking to external IMDB data, even though it has many nulls.

4. backdrop_path (779,429 nulls)
- While mostly null, it might be useful for visual representations in some applications.

5. release_date and release_year (both 150,556 nulls)
- These are important for temporal analysis, and the null count is relatively low compared to the dataset size.

6. Star1, Star2, Star3, Star4 (all have 1,041,942 or more nulls)
- **Explanation:** These are mostly null and redundant with the `Cast_list` attribute, which has no null values. But are important in determining the potential popularity of a movie based on its star-studded cast.

In [16]:
# Assuming your DataFrame is named 'imdb_df'

unnecessary_attributes = [
    'AverageRating',
    'Poster_Link',
    'Certificate',
    'IMDB_Rating',
    'Meta_score',
    'Music_Composer',
    'Director_of_Photography',
    'homepage',
    'tagline',
    'Producers',
    'Writer'
]

# Drop the unnecessary columns
df2 = df.drop(columns=unnecessary_attributes)
df2.columns

Index(['id', 'title', 'vote_average', 'vote_count', 'status', 'release_date',
       'revenue', 'runtime', 'adult', 'backdrop_path', 'budget', 'imdb_id',
       'original_language', 'original_title', 'overview', 'popularity',
       'poster_path', 'production_companies', 'production_countries',
       'spoken_languages', 'keywords', 'release_year', 'Director', 'Star1',
       'Star2', 'Star3', 'Star4', 'genres_list', 'Cast_list',
       'overview_sentiment', 'all_combined_keywords'],
      dtype='object')

In [17]:
df2.isnull().sum()

id                             0
title                          0
vote_average                   0
vote_count                     0
status                         0
release_date              150556
revenue                        0
runtime                        0
adult                          0
backdrop_path             779429
budget                         0
imdb_id                   482891
original_language              0
original_title                 0
overview                  202181
popularity                     0
poster_path               317838
production_companies      575683
production_countries      456376
spoken_languages          440257
keywords                       0
release_year              150556
Director                       0
Star1                    1041942
Star2                    1069485
Star3                    1069485
Star4                    1069485
genres_list                    0
Cast_list                      0
overview_sentiment             0
all_combin

In [18]:
df[['release_date', 'release_year']].head()

release_date  release_year
0   2010-07-15        2010.0
1   2014-11-05        2014.0
2   2008-07-16        2008.0
3   2009-12-15        2009.0
4   2012-04-25        2012.0

In [19]:
df[['release_date', 'release_year']].tail()

release_date  release_year
1072250   2017-07-13        2017.0
1072251   2019-06-12        2019.0
1072252   2020-06-15        2020.0
1072253   1964-02-02        1964.0
1072254   1921-11-27        1921.0

In [20]:
df[['release_date', 'release_year']].isnull().sum()

release_date    150556
release_year    150556
dtype: int64

In [21]:
df[(df['release_date'].isnull()) & (df['release_year'].isnull())].isnull().sum()

id                              0
title                           0
vote_average                    0
vote_count                      0
status                          0
release_date               150556
revenue                         0
runtime                         0
adult                           0
backdrop_path              141284
budget                          0
homepage                   136442
imdb_id                    109345
original_language               0
original_title                  0
overview                    48112
popularity                      0
poster_path                100523
tagline                    132378
production_companies       132145
production_countries       128879
spoken_languages           116500
keywords                        0
release_year               150556
Director                        0
AverageRating              149555
Poster_Link                150300
Certificate                147776
IMDB_Rating                147716
Meta_score    

The time data of this dataset is important to us, and thus we will not be able to use those rows without a date.

In [22]:
df2.dropna(subset=['release_date'], inplace=True)
df2.dropna(subset=['release_year'], inplace=True)
df2.shape

(921699, 31)

In [23]:
df2[['Star1', 'Star2', 'Star3', 'Star4']].head()

Star1                 Star2               Star3  \
0    Leonardo DiCaprio  Joseph Gordon-Levitt         Elliot Page   
1  Matthew McConaughey         Anne Hathaway    Jessica Chastain   
2       Christian Bale          Heath Ledger       Aaron Eckhart   
3      Sam Worthington           Zoe Saldana    Sigourney Weaver   
4    Robert Downey Jr.           Chris Evans  Scarlett Johansson   

                Star4  
0        Ken Watanabe  
1       Mackenzie Foy  
2       Michael Caine  
3  Michelle Rodriguez  
4       Jeremy Renner

In [24]:
df2[['Star1', 'Star2', 'Star3', 'Star4']].tail()

Star1 Star2 Star3 Star4
1072250   NaN   NaN   NaN   NaN
1072251   NaN   NaN   NaN   NaN
1072252   NaN   NaN   NaN   NaN
1072253   NaN   NaN   NaN   NaN
1072254   NaN   NaN   NaN   NaN

This identifies if there are any popular/famous movie starts starring in the movie. So it is ok to have `null` values.

In [25]:
df2['Star1'] = df2['Star1'].fillna('Unknown')
df2['Star2'] = df2['Star2'].fillna('Unknown')
df2['Star3'] = df2['Star3'].fillna('Unknown')
df2['Star4'] = df2['Star4'].fillna('Unknown')
df2[['Star1', 'Star2', 'Star3', 'Star4']].tail()

Star1    Star2    Star3    Star4
1072250  Unknown  Unknown  Unknown  Unknown
1072251  Unknown  Unknown  Unknown  Unknown
1072252  Unknown  Unknown  Unknown  Unknown
1072253  Unknown  Unknown  Unknown  Unknown
1072254  Unknown  Unknown  Unknown  Unknown

In [26]:
df2[['production_countries', 'production_companies']].head()

production_countries  \
0  United Kingdom, United States of America   
1  United Kingdom, United States of America   
2  United Kingdom, United States of America   
3  United States of America, United Kingdom   
4                  United States of America   

                                production_companies  
0  Legendary Pictures, Syncopy, Warner Bros. Pict...  
1  Legendary Pictures, Syncopy, Lynda Obst Produc...  
2  DC Comics, Legendary Pictures, Syncopy, Isobel...  
3  Dune Entertainment, Lightstorm Entertainment, ...  
4                                     Marvel Studios

In [27]:
df2[['production_countries', 'production_companies']].tail()

production_countries               production_companies
1072250                  NaN                                NaN
1072251               Poland                           PWSFTviT
1072252          South Korea  Korea National University of Arts
1072253               Greece                     Spentzos Films
1072254                  NaN                                NaN

In [28]:
df2[(df2['production_companies'].isnull()) & (df2['production_countries'].isnull())].isnull().sum()

id                            0
title                         0
vote_average                  0
vote_count                    0
status                        0
release_date                  0
revenue                       0
runtime                       0
adult                         0
backdrop_path            228591
budget                        0
imdb_id                  146376
original_language             0
original_title                0
overview                  50740
popularity                    0
poster_path               92545
production_companies     268835
production_countries     268835
spoken_languages         180182
keywords                      0
release_year                  0
Director                      0
Star1                         0
Star2                         0
Star3                         0
Star4                         0
genres_list                   0
Cast_list                     0
overview_sentiment            0
all_combined_keywords         0
dtype: i

In [29]:
df2['production_companies'] = df2['production_companies'].fillna('Unknown')
df2['production_countries'] = df2['production_countries'].fillna('Unknown')
df2[['production_companies', 'production_countries']].tail()

production_companies production_countries
1072250                            Unknown              Unknown
1072251                           PWSFTviT               Poland
1072252  Korea National University of Arts          South Korea
1072253                     Spentzos Films               Greece
1072254                            Unknown              Unknown

In [30]:
df2.isnull().sum()

id                            0
title                         0
vote_average                  0
vote_count                    0
status                        0
release_date                  0
revenue                       0
runtime                       0
adult                         0
backdrop_path            638145
budget                        0
imdb_id                  373546
original_language             0
original_title                0
overview                 154069
popularity                    0
poster_path              217315
production_companies          0
production_countries          0
spoken_languages         323757
keywords                      0
release_year                  0
Director                      0
Star1                         0
Star2                         0
Star3                         0
Star4                         0
genres_list                   0
Cast_list                     0
overview_sentiment            0
all_combined_keywords         0
dtype: i

In [31]:
df2[['backdrop_path', 'poster_path']].head()

backdrop_path                       poster_path
0  /8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg  /oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg
1  /pbrkL804c8yAv3zBZR4QPEafpAR.jpg  /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg
2  /nMKdUUepR0i5zn0y1T4CsSB5chy.jpg  /qJ2tW6WMUDux911r6m7haRef0WH.jpg
3  /vL5LR6WdxWPjLPFRLe133jXWsh5.jpg  /kyeqWdyUXW608qlYkRqosgbbJyK.jpg
4  /9BBTo63ANSmhC4e6r62OJFuK2GL.jpg   /RYMX2wcKCBAr24UyPD7xwmjaTn.jpg

In [32]:
df2[['backdrop_path', 'poster_path']].tail()

backdrop_path                       poster_path
1072250                               NaN  /tKnwO0Tc7p8gq1YffNwUPuXbl6f.jpg
1072251  /xfSlHmpbqOAHaWMVzyN3XXF4PyE.jpg  /uVRCFcgvyiPUqAHO1xYoeUqBss9.jpg
1072252                               NaN  /9y6hHj6gTpGIilOgN2LhNh8XHIQ.jpg
1072253  /cIKFTIJQqhb9EX1YIgPGlm2voyK.jpg  /ekLcnTMKm3rwwCfzApullRbTT4R.jpg
1072254                               NaN                               NaN

On more intricate analysis, these attributes do not contribute to the project.

In [33]:
df2 = df2.drop(columns=['backdrop_path', 'poster_path'])
df2.columns

Index(['id', 'title', 'vote_average', 'vote_count', 'status', 'release_date',
       'revenue', 'runtime', 'adult', 'budget', 'imdb_id', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'spoken_languages', 'keywords', 'release_year',
       'Director', 'Star1', 'Star2', 'Star3', 'Star4', 'genres_list',
       'Cast_list', 'overview_sentiment', 'all_combined_keywords'],
      dtype='object')

In [34]:
df2[['imdb_id', 'overview', 'spoken_languages']].head()

imdb_id                                           overview  \
0  tt1375666  Cobb, a skilled thief who commits corporate es...   
1  tt0816692  The adventures of a group of explorers who mak...   
2  tt0468569  Batman raises the stakes in his war on crime. ...   
3  tt0499549  In the 22nd century, a paraplegic Marine is di...   
4  tt0848228  When an unexpected enemy emerges and threatens...   

                     spoken_languages  
0  English, French, Japanese, Swahili  
1                             English  
2                   English, Mandarin  
3                    English, Spanish  
4             English, Hindi, Russian

In [35]:
df2[['imdb_id', 'overview', 'spoken_languages']].tail()

imdb_id                                           overview  \
1072250         NaN  Mahmoud and Nabila are married couple in their...   
1072251  tt10784416  In order to confront her unfaithful partner, a...   
1072252         NaN  A high school student. My best friend is pregn...   
1072253   tt0231143                                                NaN   
1072254   tt0403050                    An Aesop’s Film Fables cartoon.   

        spoken_languages  
1072250           Arabic  
1072251           Polish  
1072252              NaN  
1072253            Greek  
1072254              NaN

`imdb_id` is not useful for further research, while `overview` and `spoken_languages` are still useful.

In [36]:
df2 = df2.drop(columns=['imdb_id'])
df2['overview'] = df2['overview'].fillna('Unknown')
df2['spoken_languages'] = df2['spoken_languages'].fillna('Unknown')
df2.isnull().sum()

id                       0
title                    0
vote_average             0
vote_count               0
status                   0
release_date             0
revenue                  0
runtime                  0
adult                    0
budget                   0
original_language        0
original_title           0
overview                 0
popularity               0
production_companies     0
production_countries     0
spoken_languages         0
keywords                 0
release_year             0
Director                 0
Star1                    0
Star2                    0
Star3                    0
Star4                    0
genres_list              0
Cast_list                0
overview_sentiment       0
all_combined_keywords    0
dtype: int64

In [37]:
df2 = df2.drop(columns=['original_title', 'all_combined_keywords'])
df2.columns

Index(['id', 'title', 'vote_average', 'vote_count', 'status', 'release_date',
       'revenue', 'runtime', 'adult', 'budget', 'original_language',
       'overview', 'popularity', 'production_companies',
       'production_countries', 'spoken_languages', 'keywords', 'release_year',
       'Director', 'Star1', 'Star2', 'Star3', 'Star4', 'genres_list',
       'Cast_list', 'overview_sentiment'],
      dtype='object')

### Row Expansion

In [38]:
df2['stars'] = df2['Star1'] + ', ' + df2['Star2'] + ', ' + df2['Star3'] + ', ' + df2['Star4']
df2['stars'].head()

0    Leonardo DiCaprio, Joseph Gordon-Levitt, Ellio...
1    Matthew McConaughey, Anne Hathaway, Jessica Ch...
2    Christian Bale, Heath Ledger, Aaron Eckhart, M...
3    Sam Worthington, Zoe Saldana, Sigourney Weaver...
4    Robert Downey Jr., Chris Evans, Scarlett Johan...
Name: stars, dtype: object

In [39]:
df2['stars'] = df2['stars'].progress_apply(lambda x: x.split(', '))
df2['stars'].head()

  0%|          | 0/921699 [00:00<?, ?it/s]

0    [Leonardo DiCaprio, Joseph Gordon-Levitt, Elli...
1    [Matthew McConaughey, Anne Hathaway, Jessica C...
2    [Christian Bale, Heath Ledger, Aaron Eckhart, ...
3    [Sam Worthington, Zoe Saldana, Sigourney Weave...
4    [Robert Downey Jr., Chris Evans, Scarlett Joha...
Name: stars, dtype: object

In [40]:
df2 = df2.drop(columns=['Star1', 'Star2', 'Star3', 'Star4'])

In [41]:
df2.shape

(921699, 23)

In [42]:
df2.columns

Index(['id', 'title', 'vote_average', 'vote_count', 'status', 'release_date',
       'revenue', 'runtime', 'adult', 'budget', 'original_language',
       'overview', 'popularity', 'production_companies',
       'production_countries', 'spoken_languages', 'keywords', 'release_year',
       'Director', 'genres_list', 'Cast_list', 'overview_sentiment', 'stars'],
      dtype='object')

In [43]:
for i in tqdm(df2.columns):
    print(f'{i}: {type(df2[i][0])}, {df2[i][0]}')

  0%|          | 0/23 [00:00<?, ?it/s]

id: <class 'numpy.int64'>, 27205
title: <class 'str'>, Inception
vote_average: <class 'numpy.float64'>, 8.364
vote_count: <class 'numpy.int64'>, 34495
status: <class 'str'>, Released
release_date: <class 'pandas._libs.tslibs.timestamps.Timestamp'>, 2010-07-15 00:00:00
revenue: <class 'numpy.int64'>, 825532764
runtime: <class 'numpy.int64'>, 148
adult: <class 'numpy.bool_'>, False
budget: <class 'numpy.int64'>, 160000000
original_language: <class 'str'>, en
overview: <class 'str'>, Cobb, a skilled thief who commits corporate espionage by infiltrating the subconscious of his targets is offered a chance to regain his old life as payment for a task considered to be impossible: "inception", the implantation of another person's idea into a target's subconscious.
popularity: <class 'numpy.float64'>, 83.952
production_companies: <class 'str'>, Legendary Pictures, Syncopy, Warner Bros. Pictures
production_countries: <class 'str'>, United Kingdom, United States of America
spoken_languages: <clas

In [44]:
df2['production_companies'] = df2['production_companies'].progress_apply(lambda x: x.split(', '))
df2['production_countries'] = df2['production_countries'].progress_apply(lambda x: x.split(', '))
df2['spoken_languages'] = df2['spoken_languages'].progress_apply(lambda x: x.split(', '))
df2['Director'] = df2['Director'].progress_apply(lambda x: x.split(', '))
df2.head()

  0%|          | 0/921699 [00:00<?, ?it/s]

  0%|          | 0/921699 [00:00<?, ?it/s]

  0%|          | 0/921699 [00:00<?, ?it/s]

  0%|          | 0/921699 [00:00<?, ?it/s]

id            title  vote_average  vote_count    status release_date  \
0   27205        Inception         8.364       34495  Released   2010-07-15   
1  157336     Interstellar         8.417       32571  Released   2014-11-05   
2     155  The Dark Knight         8.512       30619  Released   2008-07-16   
3   19995           Avatar         7.573       29815  Released   2009-12-15   
4   24428     The Avengers         7.710       29166  Released   2012-04-25   

      revenue  runtime  adult     budget  ...  \
0   825532764      148  False  160000000  ...   
1   701729206      169  False  165000000  ...   
2  1004558444      152  False  185000000  ...   
3  2923706026      162  False  237000000  ...   
4  1518815515      143  False  220000000  ...   

                                production_companies  \
0  [Legendary Pictures, Syncopy, Warner Bros. Pic...   
1  [Legendary Pictures, Syncopy, Lynda Obst Produ...   
2  [DC Comics, Legendary Pictures, Syncopy, Isobe...   
3  [Dune Entertainment, Lightstorm Entertainment,...   
4                                   [Marvel Studios]   

                         production_countries  \
0  [United Kingdom, United States of America]   
1  [United Kingdom, United States of America]   
2  [United Kingdom, United States of America]   
3  [United States of America, United Kingdom]   
4                  [United States of America]   

                       spoken_languages  \
0  [English, French, Japanese, Swahili]   
1                             [English]   
2                   [English, Mandarin]   
3                    [English, Spanish]   
4             [English, Hindi, Russian]   

                                            keywords release_year  \
0  [rescue, mission, dream, airplane, paris, fran...       2010.0   
1  [rescue, future, spacecraft, race against time...       2014.0   
2  [joker, sadism, chaos, secret identity, crime ...       2008.0   
3  [future, society, culture clash, space travel,...       2009.0   
4  [new york city, superhero, shield, based on co...       2012.0   

              Director                                    genres_list  \
0  [Christopher Nolan]           [Action, Science Fiction, Adventure]   
1  [Christopher Nolan]            [Adventure, Drama, Science Fiction]   
2  [Christopher Nolan]               [Drama, Action, Crime, Thriller]   
3      [James Cameron]  [Action, Adventure, Fantasy, Science Fiction]   
4        [Joss Whedon]           [Science Fiction, Action, Adventure]   

                                           Cast_list overview_sentiment  \
0  [Tim Kelleher, Silvie Laguna, Natasha Beaumont...          -0.011111   
1  [Jeff Hephner, William Devane, Elyes Gabel, To...           0.045455   
2  [Tommy Lister Jr., Edison Chen, Beatrice Rosen...           0.025000   
3  [Carvon Futrell, Joel David Moore, Jon Curry, ...           0.041667   
4  [Haneyuri, Nako Mizusawa, Marin, Rikako Sakata...           0.025000   

                                               stars  
0  [Leonardo DiCaprio, Joseph Gordon-Levitt, Elli...  
1  [Matthew McConaughey, Anne Hathaway, Jessica C...  
2  [Christian Bale, Heath Ledger, Aaron Eckhart, ...  
3  [Sam Worthington, Zoe Saldana, Sigourney Weave...  
4  [Robert Downey Jr., Chris Evans, Scarlett Joha...  

[5 rows x 23 columns]

In [45]:
df2['genre'] = df2['genres_list']
df2 = df2.drop(columns=['genres_list'])
df2['Cast'] = df2['Cast_list']
df2 = df2.drop(columns=['Cast_list'])

In [46]:
df3 = df2.explode('production_companies').explode('production_countries').explode('spoken_languages').explode('Director').explode('stars').explode('genre').explode('Cast')
df3.shape

(112510316, 23)

In [47]:
df3.head()

id      title  vote_average  vote_count    status release_date  \
0  27205  Inception         8.364       34495  Released   2010-07-15   
0  27205  Inception         8.364       34495  Released   2010-07-15   
0  27205  Inception         8.364       34495  Released   2010-07-15   
0  27205  Inception         8.364       34495  Released   2010-07-15   
0  27205  Inception         8.364       34495  Released   2010-07-15   

     revenue  runtime  adult     budget  ... production_companies  \
0  825532764      148  False  160000000  ...   Legendary Pictures   
0  825532764      148  False  160000000  ...   Legendary Pictures   
0  825532764      148  False  160000000  ...   Legendary Pictures   
0  825532764      148  False  160000000  ...   Legendary Pictures   
0  825532764      148  False  160000000  ...   Legendary Pictures   

  production_countries  spoken_languages  \
0       United Kingdom           English   
0       United Kingdom           English   
0       United Kingdom           English   
0       United Kingdom           English   
0       United Kingdom           English   

                                            keywords release_year  \
0  [rescue, mission, dream, airplane, paris, fran...       2010.0   
0  [rescue, mission, dream, airplane, paris, fran...       2010.0   
0  [rescue, mission, dream, airplane, paris, fran...       2010.0   
0  [rescue, mission, dream, airplane, paris, fran...       2010.0   
0  [rescue, mission, dream, airplane, paris, fran...       2010.0   

            Director overview_sentiment              stars   genre  \
0  Christopher Nolan          -0.011111  Leonardo DiCaprio  Action   
0  Christopher Nolan          -0.011111  Leonardo DiCaprio  Action   
0  Christopher Nolan          -0.011111  Leonardo DiCaprio  Action   
0  Christopher Nolan          -0.011111  Leonardo DiCaprio  Action   
0  Christopher Nolan          -0.011111  Leonardo DiCaprio  Action   

               Cast  
0      Tim Kelleher  
0     Silvie Laguna  
0  Natasha Beaumont  
0    Kraig Thornber  
0       Jack Murray  

[5 rows x 23 columns]

### Export Data

In [ ]:
df3.to_csv('../data/imdb_imdb_cleaned.csv', index=False)

In [48]:
# # Open the file in write mode
# with open('../data/imdb_tmdb_cleaned.csv', 'w', newline='', encoding='utf-8') as f:
#     # Write the header first
#     df3.iloc[:0].to_csv(f, index=False)
    
#     # Use tqdm to track the progress of writing rows to CSV
#     for i in tqdm(range(0, len(df3))):
#         df3.iloc[[i]].to_csv(f, header=False, index=False)

  0%|          | 0/112510316 [00:00<?, ?it/s]